README.rst
=========
Mopytools
=========

Mopytools provides command-line tools to build Python application that follow
Mozilla Services standard:

- buildapp: builds an application inplace.
- buildrpms: builds a collection of RPMs for the project


Those scripts are usually driven from a project's Makefile.



CHANGES.rst
Changelog
---------

3.4 - 2014-01-03
================

- don't pass an empty --extra-index-url through to pip.


3.3 - 2012-06-29
================

- improve support for building git-hosted dependencies.
- error out if there are unpined dependencies in stage/prod reqs file.


3.2 - 2012-03-29
================

- make sure we can build without reqs files.

3.1 - 2012-03-26
================

- now allow internal dependencies to be without .spec

3.0 - 2012-03-23
================

- make sure we can use full urls with deps

2.9 - 2012-03-13
================

- pinned Distutils 2


2.8 - 2012-02-12
================

- skip lines that starts with '#' in requirement files
- added an option so we may exit the process immediatly
  when the exit code is not 0


2.7 - 2012-01-30
================

- tagging again - nothing new


2.6 - 2012-01-30
================

- make sure build_rpm checks the force option.


2.5 - 2012-01-30
================

- Broken release


2.4 - 2012-01-25
================

- sort release tags by version # instead of default (i.e. order of creation)

2.3 - 2012-01-17
================

- Fixed behavior on systems w/ hg diff tool installed
  (i.e. use 'hg diff' instead of 'hg di')
- Added support for TAG_PREFIX specified by an environment variable

2.2 - 2011-12-11
================

- activate the download cache on pypi2rpm

2.1 - 2011-12-11
================

- added the download cache option

2.0 - 2011-12-2
===============

- Make sure Git Submodules get updated

1.9 - 2011-11-30
================

- Raised the default timeout to 300s
- Make sure Pip uses the index/extras options

1.8 - 2011-11-24
================

- Make sure the req building occurs in a clean env,


1.7 - 2011-11-14
================

- Bug 700242 - add a verbose and timeout option


1.6 - 2011-11-03
================

- Bug 695232 - fixed the pinning


1.3 to 1.5 - 2011-10-06
=======================

- fixed small git-specific issues

1.2 - 2011-10-06
================

- Make the tool work with git.


1.1 - 2011-09-21
================

- use default and not tip when building

1.0 - 2011-09-09
================

- make sure we raise an error on bad number of arguments
- added a -r option to remove the destination dir for rpms


0.8 - 2011-08-19
================

- provide hgrc as a fallback


0.6 - 2011-08-16
================

- make sure the rpm trailing version is taken into account


0.5 - 2011-08-16
================

- added a buildrpm script
- new behavior for the buildapp script


0.4
===

- The build script now supports three options:

  -i: PyPI Simple index location (for mirroring)
  -e: extra location (for archives not present at PyPI)
  -s: flag to indicate that any URL that's not under PyPI or the extra
      location is disallowed during the fetching





mopytools/tests/test_build.py
# ***** BEGIN LICENSE BLOCK *****
# Version: MPL 1.1/GPL 2.0/LGPL 2.1
#
# The contents of this file are subject to the Mozilla Public License Version
# 1.1 (the "License"); you may not use this file except in compliance with
# the License. You may obtain a copy of the License at
# http://www.mozilla.org/MPL/
#
# Software distributed under the License is distributed on an "AS IS" basis,
# WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
# for the specific language governing rights and limitations under the
# License
#
# The Original Code is Sync Server
#
# The Initial Developer of the Original Code is the Mozilla Foundation.
# Portions created by the Initial Developer are Copyright (C) 2010
# the Initial Developer. All Rights Reserved.
#
# Contributor(s):
#   Tarek Ziade (tarek@mozilla.com)
#
# Alternatively, the contents of this file may be used under the terms of
# either the GNU General Public License Version 2 or later (the "GPL"), or
# the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
# in which case the provisions of the GPL or the LGPL are applicable instea
# of those above. If you wish to allow use of your version of this file only
# under the terms of either the GPL or the LGPL, and not to allow others to
# use your version of this file under the terms of the MPL, indicate your
# decision by deleting the provisions above and replace them with the notice
# and other provisions required by the GPL or the LGPL. If you do not delete
# the provisions above, a recipient may use your version of this file under
# the terms of any one of the MPL, the GPL or the LGPL.
#
# ***** END LICENSE BLOCK *****
""" tests for mopytools.build
"""
import unittest
import subprocess
import StringIO
import ConfigParser
import sys
import tempfile
import os

from mopytools.util import get_channel_tag, tag_exists, get_options
from mopytools import util


_CMDS = {"hg tags": """\
tip                               35:7d3a88af29ec
rpm-0.5rc1                        33:51e4cfb38a04
rpm-0.4                           32:c56849d09a4c
rpm-0.4rc2                        32:51e4cfb38a04
rpm-0.4rc1                        32:51e4cfb38a04
rpm-0.3                           28:51e4cfb38a04
rpm-0.2                           9:d6f665b7d6a3""",

        "git tag": """\
rpm-0.2
rpm-0.3
rpm-0.4rc1
rpm-0.4rc2
rpm-0.4
rpm-0.5rc1""",
         "git branch": """\
* master
  feature""",

         "hg branches": """\
default                           35:7d3a88af29ec
feature                           34:fca8887c0991"""}


_CFG = """\
[easy_install]
index_url = http://pypi.python.org/simple

"""


class FakePopen(object):
    def __init__(self, command, *args, **kw):
        self.cmd = command
        self.stdout = StringIO.StringIO(_CMDS[command])


class ParserNoWrite(ConfigParser.ConfigParser):
    writes = []

    def write(self, cfg):
        stream = StringIO.StringIO()
        ConfigParser.ConfigParser.write(self, stream)
        stream.seek(0)
        self.writes.append((cfg, stream.read()))


class TestBuild(unittest.TestCase):

    def setUp(self):
        self.old_po = subprocess.Popen
        subprocess.Popen = FakePopen
        self.old_cp = util.ConfigParser
        util.ConfigParser = ParserNoWrite

    def tearDown(self):
        subprocess.Popen = self.old_po
        util.ConfigParser = self.old_cp

    def test_get_tag(self):
        self.assertEqual(get_channel_tag('dev'), 'master')
        self.assertEqual(get_channel_tag('prod'), 'rpm-0.4')
        self.assertEqual(get_channel_tag('stage'), 'rpm-0.5rc1')

    def test_get_tag_actually_branch(self):
        self.assertFalse('feature' in util._get_tags())
        self.assertTrue('feature' in util._get_tags(prefix=''))

    def test_tag_exists(self):
        self.assertTrue(tag_exists('tip'))
        self.assertTrue(tag_exists('rpm-0.4'))
        self.assertTrue(tag_exists('rpm-0.5rc1'))
        self.assertFalse(tag_exists('xxx'))

    def test_distutils_setup(self):
        old_argv = sys.argv[:]
        sys.argv[:] = ['whatever', 'is_done']
        try:
            options, args = get_options()
        finally:
            sys.argv[:] = old_argv

        result = ParserNoWrite.writes[-1][1]
        self.assertEquals(result, _CFG)

    def test_rmdir(self):
        from mopytools import build_rpms

        old_build_app = build_rpms.build_app
        build_rpms.build_app = lambda: None
        old_buildrpms = build_rpms._buildrpms
        build_rpms._buildrpms = lambda deps, channel, options: None
        old_argv = sys.argv[:]

        tempdir = tempfile.mkdtemp()
        with open(os.path.join(tempdir, 'xx'), 'w') as f:
            f.write('#')

        sys.argv[:] = ['', '-r', '--dist-dir', tempdir]
        old_stdout = sys.stdout
        sys.stdout = StringIO.StringIO()
        try:
            build_rpms.main()
        finally:
            sys.argv[:] = old_argv
            build_rpms.build_app = old_build_app
            build_rpms._buildrpms = old_buildrpms
            sys.stdout = old_stdout

        try:
            # let's check that the dir was removed and recreated empty
            self.assertTrue(os.path.exists(tempdir))
            self.assertEquals(len(os.listdir(tempdir)), 0)
        finally:
            os.rmdir(tempdir)

    def test_stabby(self):
        old_argv = sys.argv[:]
        sys.argv[:] = ['', 'one, ', 'two']
        old_stdout = sys.stdout
        sys.stdout = StringIO.StringIO()
        try:
            self.assertRaises(SystemExit, get_options)
        finally:
            sys.argv[:] = old_argv
            sys.stdout = old_stdout

    def test_pinning(self):
        reqs = os.path.join(os.path.dirname(__file__), 'dev-reqs.txt')
        unpinned = util.get_non_pinned(reqs)
        unpinned.sort()
        self.assertEqual(unpinned,
                         ['Paste', 'translationstring', 'wsgi-intercept'])


mopytools/build.py
# ***** BEGIN LICENSE BLOCK *****
# Version: MPL 1.1/GPL 2.0/LGPL 2.1
#
# The contents of this file are subject to the Mozilla Public License Version
# 1.1 (the "License"); you may not use this file except in compliance with
# the License. You may obtain a copy of the License at
# http://www.mozilla.org/MPL/
#
# Software distributed under the License is distributed on an "AS IS" basis,
# WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
# for the specific language governing rights and limitations under the
# License.
#
# The Original Code is Sync Server
#
# The Initial Developer of the Original Code is the Mozilla Foundation.
# Portions created by the Initial Developer are Copyright (C) 2010
# the Initial Developer. All Rights Reserved.
#
# Contributor(s):
#   Tarek Ziade (tarek@mozilla.com)
#
# Alternatively, the contents of this file may be used under the terms of
# either the GNU General Public License Version 2 or later (the "GPL"), or
# the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
# in which case the provisions of the GPL or the LGPL are applicable instead
# of those above. If you wish to allow use of your version of this file only
# under the terms of either the GPL or the LGPL, and not to allow others to
# use your version of this file under the terms of the MPL, indicate your
# decision by deleting the provisions above and replace them with the notice
# and other provisions required by the GPL or the LGPL. If you do not delete
# the provisions above, a recipient may use your version of this file under
# the terms of any one of the MPL, the GPL or the LGPL.
#
# ***** END LICENSE BLOCK *****
import os
import sys

from mopytools.util import (run, envname, update_cmd, step,
                            get_project_name, is_meta_project,
                            has_changes, is_git)


@step('Updating the repo')
def updating_repo(name, channel, specific_tags, force=False, timeout=60,
                  verbose=False):
    if not force and has_changes(timeout, verbose) and channel != 'dev':
        print('The code was changed locally, aborting!')
        print('You can use --force but all uncommited '
              'changes will be discarded.')
        sys.exit(0)

    if is_git():
        run('git submodule update')

    run(update_cmd(name, channel, specific_tags, force), timeout, verbose)


@step('Checking provided tags')
def check_tags(projects):
    tags = {}
    missing = provided = 0
    for project in projects:
        tag = envname(project)
        if tag in os.environ:
            tags[tag] = os.environ[tag]
            provided += 1
        else:
            tags[tag] = 'Not provided'
            missing += 1

    # we want all tag or no tag
    if missing > 0 and missing < len(projects):
        print("You did not specify all tags: ")
        for project, tag in tags.items():
            print('    %s: %s' % (project, tag))

        print("Also, consider using channels")
        sys.exit(1)

    return provided == len(projects) and missing == 0


@step('Checking the environ')
def get_environ_info(deps):
    name = get_project_name()
    # is the root a project itself or just a placeholder ?
    projects = list(deps)
    if not is_meta_project():
        projects.append(name)

    # check the tags
    tags = {}
    missing = provided = 0
    for project in projects:
        tag = envname(project)
        if tag in os.environ:
            tags[tag] = os.environ[tag]
            provided += 1
        else:
            tags[tag] = 'Not provided'
            missing += 1

    # we want all tag or no tag
    if missing > 0 and missing < len(projects):
        print("You did not specify all tags: ")
        for project, tag in tags.items():
            print('    %s: %s' % (project, tag))

        print("Also, consider using channels")
        sys.exit(1)

    specific_tags = provided == len(projects) and missing == 0
    return name, specific_tags


mopytools/buildsrv.py
# ***** BEGIN LICENSE BLOCK *****
# Version: MPL 1.1/GPL 2.0/LGPL 2.1
#
# The contents of this file are subject to the Mozilla Public License Version
# 1.1 (the "License"); you may not use this file except in compliance with
# the License. You may obtain a copy of the License at
# http://www.mozilla.org/MPL/
#
# Software distributed under the License is distributed on an "AS IS" basis,
# WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
# for the specific language governing rights and limitations under the
# License.
#
# The Original Code is Sync Server
#
# The Initial Developer of the Original Code is the Mozilla Foundation.
# Portions created by the Initial Developer are Copyright (C) 2010
# the Initial Developer. All Rights Reserved.
#
# Contributor(s):
#   Tarek Ziade (tarek@mozilla.com)
#
# Alternatively, the contents of this file may be used under the terms of
# either the GNU General Public License Version 2 or later (the "GPL"), or
# the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
# in which case the provisions of the GPL or the LGPL are applicable instead
# of those above. If you wish to allow use of your version of this file only
# under the terms of either the GPL or the LGPL, and not to allow others to
# use your version of this file under the terms of the MPL, indicate your
# decision by deleting the provisions above and replace them with the notice
# and other provisions required by the GPL or the LGPL. If you do not delete
# the provisions above, a recipient may use your version of this file under
# the terms of any one of the MPL, the GPL or the LGPL.
#
# ***** END LICENSE BLOCK *****
"""
Bootstrap file server
"""
import os
import mimetypes
from hashlib import md5
from wsgiref.simple_server import make_server

# where the served files are located ?
_DIR = '.'


def application(environ, start_response):
    if_none_match = environ.get('HTTP_IF_NONE_MATCH', '')
    path = environ['PATH_INFO']
    path = path.lstrip('/')

    if path.startswith('.') or '/' in path:
        # looks like a bad path -- or a potential security thread
        status = '400 Bad Request'
        headers = [('Content-type', 'text/plain')]
        start_response(status, headers)
        return "Invalid Request"

    path = os.path.join(_DIR, path)

    if not os.path.isfile(path):
        # unknown file
        status = '404 Not Found'
        headers = [('Content-type', 'text/plain')]
        start_response(status, headers)
        return "File not found"

    # we're good, let's look at the file age
    etag = '"%s"' % md5(str(os.stat(path).st_mtime)).hexdigest()
    if etag == if_none_match:
        # same file
        status = '412 Precondition failed'
        headers = [('Content-type', 'text/plain')]
        start_response(status, headers)
        return ''

    content_type = mimetypes.guess_type(path)
    status = '200 OK'
    headers = [('Content-type', content_type[0]),
               ('ETag', etag)]

    start_response(status, headers)

    with open(path) as f:
        content = f.read()

    return content


if __name__ == '__main__':
    httpd = make_server('', 5000, application)
    print "Listening on port 5000...."
    httpd.serve_forever()


mopytools/apptemplate/__init__.py
# ***** BEGIN LICENSE BLOCK *****
# Version: MPL 1.1/GPL 2.0/LGPL 2.1
#
# The contents of this file are subject to the Mozilla Public License Version
# 1.1 (the "License"); you may not use this file except in compliance with
# the License. You may obtain a copy of the License at
# http://www.mozilla.org/MPL/
#
# Software distributed under the License is distributed on an "AS IS" basis,
# WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
# for the specific language governing rights and limitations under the
# License.
#
# The Original Code is Sync Server
#
# The Initial Developer of the Original Code is the Mozilla Foundation.
# Portions created by the Initial Developer are Copyright (C) 2010
# the Initial Developer. All Rights Reserved.
#
# Contributor(s):
#   Tarek Ziade (tarek@mozilla.com)
#
# Alternatively, the contents of this file may be used under the terms of
# either the GNU General Public License Version 2 or later (the "GPL"), or
# the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
# in which case the provisions of the GPL or the LGPL are applicable instead
# of those above. If you wish to allow use of your version of this file only
# under the terms of either the GPL or the LGPL, and not to allow others to
# use your version of this file under the terms of the MPL, indicate your
# decision by deleting the provisions above and replace them with the notice
# and other provisions required by the GPL or the LGPL. If you do not delete
# the provisions above, a recipient may use your version of this file under
# the terms of any one of the MPL, the GPL or the LGPL.
#
# ***** END LICENSE BLOCK *****
from paste.script.templates import Template, var

vars = [
    var('appname', 'Application repository name (like "server-core")'),
    var('version', 'Version (like 0.1)'),
    var('description', 'One-line description of the project'),
    var('author', 'Author name'),
    var('author_email', 'Author email'),
    var('url', 'URL of homepage (or Repository root)')]


class AppTemplate(Template):

    _template_dir = 'services_base'
    summary = "A Mozilla Services application"
    vars = vars

    def post(self, command, output_dir, vars):
        if command.verbose:
            print('Generating Application...')


mopytools/util.py
# ***** BEGIN LICENSE BLOCK *****
# Version: MPL 1.1/GPL 2.0/LGPL 2.1
#
# The contents of this file are subject to the Mozilla Public License Version
# 1.1 (the "License"); you may not use this file except in compliance with
# the License. You may obtain a copy of the License at
# http://www.mozilla.org/MPL/
#
# Software distributed under the License is distributed on an "AS IS" basis,
# WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
# for the specific language governing rights and limitations under the
# License.
#
# The Original Code is Sync Server
#
# The Initial Developer of the Original Code is the Mozilla Foundation.
# Portions created by the Initial Developer are Copyright (C) 2010
# the Initial Developer. All Rights Reserved.
#
# Contributor(s):
#   Tarek Ziade (tarek@mozilla.com)
#   Richard Newman (rnewman@mozilla.com)
#   Rob Miller (rmiller@mozilla.com)
#
# Alternatively, the contents of this file may be used under the terms of
# either the GNU General Public License Version 2 or later (the "GPL"), or
# the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
# in which case the provisions of the GPL or the LGPL are applicable instead
# of those above. If you wish to allow use of your version of this file only
# under the terms of either the GPL or the LGPL, and not to allow others to
# use your version of this file under the terms of the MPL, indicate your
# decision by deleting the provisions above and replace them with the notice
# and other provisions required by the GPL or the LGPL. If you do not delete
# the provisions above, a recipient may use your version of this file under
# the terms of any one of the MPL, the GPL or the LGPL.
#
# ***** END LICENSE BLOCK *****
import os
import sys
import re
import subprocess
import socket
from urlparse import urlparse
from ConfigParser import ConfigParser
from optparse import OptionParser
import signal

from distutils2.version import (NormalizedVersion, IrrationalVersionError,
                                suggest_normalized_version)
from distutils2.index.simple import Crawler, DEFAULT_SIMPLE_INDEX_URL
from pkg_resources import parse_version

from pip.req import parse_requirements


REPO_ROOT = 'https://hg.mozilla.org/services/'
PYTHON = sys.executable
PIP = os.path.join(os.path.dirname(PYTHON), 'pip')
PYPI = 'http://pypi.python.org/simple'
TAG_PREFIX = os.environ.get('TAG_PREFIX', 'rpm-')
PYPI2RPM = os.path.join(os.path.dirname(PYTHON), 'pypi2rpm.py')


def is_git():
    """Returns True if the current dir contains a .git"""
    return '.git' in os.listdir('.')


def _sort_tags(tags):
    # tags may come unordered
    def _sort_version(version1, version2):
        nv1 = suggest_normalized_version(version1)
        nv2 = suggest_normalized_version(version2)
        if nv1 is not None and nv2 is not None:
            # we can use d2 to sort
            if nv1 < nv2:
                return -1
            elif nv1 > nv2:
                return 1
            return 0
        else:
            # we fallback to setuptools sorting
            return cmp(parse_version(version1), parse_version(version2))

    tags.sort(cmp=_sort_version)
    tags.reverse()


def _get_tags(prefix=TAG_PREFIX):
    if is_git():
        cmds = ['git tag', 'git branch']
    else:
        cmds = ['hg tags', 'hg branches']

    output = ''
    for cmd in cmds:
        sub = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE)
        addl = sub.stdout.read().strip()
        if addl:
            if output:
                output += '\n'
            output += addl
    if output == '':
        return []

    tags = []
    for line in output.split('\n'):
        tag = line.split()[0]
        tag = tag.lstrip('*')
        tag = tag.strip()
        if tag.startswith(prefix):
            tags.append(tag)
    _sort_tags(tags)
    return tags


def tag_exists(tag):
    if tag in ('tip', 'defaut') or tag.isdigit():
        return True
    return tag in _get_tags()


def get_channel_tag(channel):
    if channel == 'dev':
        if is_git():
            return 'master'
        return 'default'

    tags = _get_tags()

    if len(tags) == 0:
        print ('Could not find any %s* tag' % TAG_PREFIX)
        sys.exit(0)

    # looking for the latest channel tag
    #
    # - prod tags are final tags
    # - stage tags is the latest rc tags that is
    #   after the latest prod tag if any, or prod tag
    def is_prod(tag):
        try:
            return NormalizedVersion(version).is_final
        except IrrationalVersionError:
            return False

    def is_stage_or_prod(tag):
        try:
            NormalizedVersion(version)
            return True
        except IrrationalVersionError:
            return False

    if channel == "prod":
        selector = is_prod
    else:
        selector = is_stage_or_prod

    for tag in tags:
        version = tag[len(TAG_PREFIX):]
        if '-' in version:
            version = version.replace('-', '.')
        if selector(version):
            return tag

    print('Could not find a tag for channel %s' % channel)
    print('Make sure you have a %s-reqs.txt file' % channel)
    sys.exit(0)


class TimeoutError(Exception):
    pass


def timeout_handler(signum, frame):
    raise TimeoutError("The call has taken too long")


def with_timer(duration, cleanup=None):
    def _timer(func):
        def __timer(*args, **kw):
            previous_sig = signal.signal(signal.SIGALRM, timeout_handler)
            signal.alarm(duration)
            try:
                return func(*args, **kw)
            except TimeoutError:
                if cleanup is not None:
                    cleanup()
                raise
            finally:
                # replace the previous sig
                signal.signal(signal.SIGALRM, previous_sig)
        return __timer
    return _timer


def run(command, timeout=300, verbose=False, allow_exit=False):
    err_output = []
    out_output = []

    def _output():
        out = 'Output:\n:%s' % '\n'.join(out_output)
        err = '\n\nErrors:\n:%s' % '\n'.join(err_output)
        return out + err

    @with_timer(timeout)
    def _run():
        if verbose:
            print('\n' + command)

        sb = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE,
                              stderr=subprocess.PIPE)

        stdout, stderr = sb.communicate()
        code = sb.returncode

        if code != 0:
            if not allow_exit or verbose:
                print("%r failed with code %d" % (command, code))
                print(stdout)
                print(stderr)
            if not allow_exit:
                sys.exit(code)
        elif verbose:
            print(stdout)
            print(stderr)

        return code, stdout, stderr

    try:
        return  _run()
    except TimeoutError:
        print(command)
        print("Timed out!")
        sys.exit(0)


def envname(name):
    return os.path.basename(name).upper().replace('-', '_')


def has_changes(timeout=5, verbose=False):
    if is_git():
        code, out, err = run('git diff --exit-code', timeout, verbose,
                             allow_exit=True)
        return code == 1
    else:
        code, out, err = run('hg diff', timeout, verbose)
        return out != ''


def update_cmd(project=None, channel="prod", specific_tag=False,
               force=False):
    if force and channel != 'dev':
        if is_git():
            cmd = 'git checkout --force'
        else:
            cmd = 'hg up -C'

    elif channel != 'dev':
        if is_git():
            cmd = 'git checkout'
        else:
            cmd = 'hg up -c'
    else:
        if is_git():
            cmd = 'git checkout'
        else:
            cmd = 'hg up'

    if not specific_tag:
        if is_git():
            return '%s "%s"' % (cmd, get_channel_tag(channel))
        else:
            return '%s -r "%s"' % (cmd, get_channel_tag(channel))

    # looking for an environ with a specific tag or rev
    if project is not None:
        rev = os.environ.get(envname(project))
        if rev is not None:
            if not tag_exists(rev):
                print('Unknown tag or revision: %s' % rev)
                sys.exit(1)

            if is_git():
                return '%s "%s"' % (cmd, rev)
            else:
                return '%s -r "%s"' % (cmd, rev)

    return cmd


_LEVEL = -1


def step(text):
    def _step(func):
        def __step(*args, **kw):
            global _LEVEL
            _LEVEL += 1
            fstep = step = _LEVEL * 2 * " "
            if _LEVEL > 0:
                fstep = '\n' + step
            msg = fstep + text % kw + ' '
            msg_len = len(msg)
            sys.stdout.write(msg)
            sys.stdout.flush()
            try:
                res = func(*args, **kw)
                pad = 100 - (len(step) + msg_len)
                msg = ('%s%' + str(pad) + 's')

                if _LEVEL > 0:
                    sys.stdout.write(msg % (step, '[ok]'))
                else:
                    sys.stdout.write(step + '\n %100s' % '[done]')
                sys.stdout.flush()
                return res
            except:
                sys.stdout.write(' %100s' % '[fail]')
                sys.stdout.flush()
                raise
            finally:
                _LEVEL -= 1
                if _LEVEL < 0:
                    print('')
        return __step
    return _step

_URL = re.compile('^Url: (.*?)$', re.M | re.DOTALL)


def is_meta_project():
    return get_spec_file() is None


def get_spec_file():
    for file_ in os.listdir(os.getcwd()):
        if os.path.splitext(file_)[-1] == '.spec':
            return os.path.join(os.getcwd(), file_)
    return None


_V = NormalizedVersion


def _match(version, token, other):
    token = token.strip()
    try:
        if token == '==':
            return version == other
        elif token == '>=':
            return _V(version) >= _V(other)
        elif token == '>':
            return _V(version) > _V(other)
        elif token == '<=':
            return _V(version) <= _V(other)
        elif token == '<':
            return _V(version) < _V(other)
        elif token == '!=':
            return _V(version) != _V(other)
    except IrrationalVersionError:
        return False
    raise NotImplementedError(token)


def _vsort(version1, version2):
    if _V(version1) > _V(version2):
        return -1
    elif _V(version1) < _V(version2):
        return 1
    return 0


def _best_release(project_name, version=None, token='==',
                  index_url=DEFAULT_SIMPLE_INDEX_URL):
    c = Crawler(index_url=index_url, prefer_final=True)
    project = c.get_releases(project_name)
    versions = project.get_versions()

    if version is None:
        versions.sort(_vsort)
        return versions[0]
    else:
        selected = []
        for existing_ver in project.get_versions():
            if _match(existing_ver, token, version):
                selected.append(existing_ver)

        if len(selected) == 0:
            print 'Unknown version'
            return None

        selected.sort(_vsort)
        return selected[0]


def split_version(line):
    tokens = ['==', '>=', '<=', '>', '<', '!=']
    for token in tokens:
        if token in line:
            if token != '==':
                # we need to grab a list of versions from
                # PyPI and decide which one works
                app, version = line.split(token, 1)
                app, version = app.strip(), version.strip()
                version = _best_release(app, version, token.strip())
            else:
                app, version = line.split(token, 1)

            return app.strip(), version.strip()
    return line.strip(), None


def get_project_name():
    if is_meta_project():
        return None

    spec_file = get_spec_file()
    if spec_file is not None:
        with open(spec_file) as f:
            data = f.read()
            name = _URL.findall(data)
            if len(name) == 1:
                return name[0].split('/')[-1]

    hgrc = os.path.join('.hg', 'hgrc')
    if os.path.exists(hgrc):
        reader = ConfigParser()
        reader.read(hgrc)
        if reader.has_option('paths', 'default'):
            repo_url = reader.get('paths', 'default')
            return repo_url.split('/')[-1]

    raise IOError('Unable to find the project name.')


def setup_pypi(pypi, extras=None, strict=False):
    # setup distutils.cfg
    import distutils
    location = os.path.dirname(distutils.__file__)
    cfg = os.path.join(location, 'distutils.cfg')
    if os.path.exists(cfg):
        os.remove(cfg)

    parser = ConfigParser()
    parser.read(cfg)

    if 'easy_install' not in parser.sections():
        parser.add_section('easy_install')

    parser.set('easy_install', 'index_url', pypi)
    allowed_hosts = [urlparse(pypi)[1]]

    if extras:
        parser.set('easy_install', 'find_links', extras)
        allowed_hosts.append(urlparse(extras)[1])

    if strict:
        parser.set('easy_install', 'allow_hosts', ','.join(allowed_hosts))

    with open(cfg, 'w') as cfg:
        parser.write(cfg)


def timeout(duration):
    def _timeout(func):
        def __timeout(*args, **kw):
            old = socket.getdefaulttimeout()
            socket.setdefaulttimeout(duration)
            try:
                return func(*args, **kw)
            finally:
                socket.setdefaulttimeout(old)
        return __timeout
    return _timeout


def get_options(extra_options=None):
    if extra_options is None:
        extra_options = []

    usage = ("usage: %prog [options] "
             "[Mozilla Services deps separated by commas]")
    parser = OptionParser(usage=usage)
    parser.add_option("-i", "--index", dest="index",
                      help="Pypi index", default=PYPI)

    parser.add_option("-e", "--extras", dest="extras",
                      help="Extra location for packages",
                      default=None)

    parser.add_option("-s", "--strict-index", dest="strict",
                      action="store_true",
                      help="Prevent browsing external websites",
                      default=False)

    parser.add_option("-c", "--channel", dest="channel",
                      help="Channel to build",
                      default="last", type="choice",
                      choices=["prod", "dev", "stage", "last"])

    parser.add_option("-f", "--force", dest="force",
                      action="store_true", default=False,
                      help="Forces update")

    parser.add_option("-v", "--verbose", dest="verbose",
                      action="store_true", default=False,
                      help="Verbose mode")

    parser.add_option("-t", "--timeout", dest="timeout",
                      help="Time out",
                      default=300, type="int")

    parser.add_option("--download-cache", dest="download_cache",
                      help="Download cache",
                      default=None)

    for optargs, optkw in extra_options:
        parser.add_option(*optargs, **optkw)

    options, args = parser.parse_args()

    if len(args) > 1:
        print('Wrong number of arguments.')
        parser.print_usage()
        sys.exit(0)

    # set pypi location
    setup_pypi(options.index, options.extras, options.strict)

    return options, args


def get_channel(options):
    channel = options.channel.lower()
    if channel == 'last':
        channel = get_last_channel()

    save_last_channel(channel)
    return channel


def save_last_channel(channel):
    with open('.channel', 'w') as f:
        f.write(channel)


def get_last_channel():
    if not os.path.exists('.channel'):
        return 'prod'
    with open('.channel') as f:
        return f.read()


class DependencyError(Exception):
    pass


class Options:
    skip_requirements_regex = False


def get_non_pinned(reqfile):
    """Return non-pinned or dev versions.

    Look for XXX==XXX patterns
    """
    res = []
    for req in parse_requirements(reqfile, options=Options()):
        if list(req.absolute_versions) == []:
            res.append(req.name)
    return res
